### PR TITLE
[MBUILDCACHE-22] added flag to force rebuild per project or globally

### PR DESCRIPTION
**Pull request by eltsovalex**
added possibility to skip generated sources restore on per-project level
added some logging

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by eltsovalex** [comment]
Sorry, I do not have an MNG number yet to add as a proper commit comment
But basically this PR is about adding 2 new features which are required to use this extension in yet another complex multi-module project in Deutsche Bank
1) ability to skip cache lookup for a particular module or whole project (required to get some modules always built e.g. via a profile - e.g. I have some additional artifacts published by CI). Alternatively it allows an easy "force" rebuild of a whole project 
The difference with not using cache is that results are put into cache, just current matching versions are not taken from caches

2) ability to disable generated source restoration on a per-module level. This is required to overcome issues with some weird Maven plugins which go nuts when cached generated sources are present and produce incorrect build result

**Comment by maximilian-novikov-db** [comment]
@eltsovalex https://issues.apache.org/jira/projects/MBUILDCACHE for this project

**Review by maximilian-novikov-db — APPROVED** [review]

**Comment by hboutemy** [comment]
is it possible to add a note in documentation about this feature and how to use it, please?
https://github.com/apache/maven-build-cache-extension/tree/master/src/site/markdown

**Review by gnodet — CHANGES_REQUESTED** [review]
Looks good, please fix the typo in the javadoc in `CacheConfigImpl.java`.

File: src/main/java/org/apache/maven/buildcache/xml/CacheConfigImpl.java [hunk 1]
```diff
@@ -87,6 +87,16 @@ public class CacheConfigImpl implements org.apache.maven.buildcache.xml.CacheCon
     public static final String RESTORE_GENERATED_SOURCES_PROPERTY_NAME = "maven.build.cache.restoreGeneratedSources";
     public static final String ALWAYS_RUN_PLUGINS = "maven.build.cache.alwaysRunPlugins";
 
+    /**
+     * Flag to control if we should skip lookup for cached artifacts globally r for a particular project even if
```
[the remark this answers is not in the document]
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/xml/CacheConfigImpl.java]: fixed

**Comment by eltsovalex** [comment]
Typos fixed in javadoc + new parameters described in documentation

**Review by maximilian-novikov-db — APPROVED** [review]

File: src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java [hunk 2]
```diff
@@ -104,9 +106,9 @@ public void execute( List<MojoExecution> mojoExecutions,
                 mojoExecutionRunner.run( mojoExecution );
             }
             cacheState = cacheConfig.initialize();
-            if ( cacheState == INITIALIZED )
+            if ( cacheState == INITIALIZED || skipCacheLookup )
             {
-                result = cacheController.findCachedBuild( session, project, mojoExecutions );
+                result = cacheController.findCachedBuild( session, project, mojoExecutions, skipCacheLookup );
```
Comment (AlexanderAshitkin) [inline src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:111]: if (skipLookup) lookup :-) 
I expected config just to return `DISABLED` state for this module

File: src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java [hunk 3]
```diff
@@ -96,6 +97,7 @@ public void execute( List<MojoExecution> mojoExecutions,
         // execute clean bound goals before restoring to not interfere/slowdown clean
         CacheState cacheState = DISABLED;
         CacheResult result = CacheResult.empty();
+        boolean skipCacheLookup = cacheConfig.isSkipLookup() || MavenProjectInput.shouldSkipCacheLookup( project );
```
Comment (AlexanderAshitkin) [inline src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java:100]: regarding the naming - lets call it skipCache. Gradle has similar parameter to force rebuild named `--no-build-cache`. from command line and pom perspective -DskipBuildCache is more descriptive

File: src/main/java/org/apache/maven/buildcache/DefaultNormalizedModelProvider.java [hunk 4]
```diff
@@ -123,10 +127,16 @@ private List<Plugin> normalizePlugins( List<Plugin> plugins )
                 {
                     Plugin copy = plugin.clone();
                     List<String> excludeProperties = cacheConfig.getEffectivePomExcludeProperties( copy );
-                    removeBlacklistedAttributes( copy.getConfiguration(), excludeProperties );
```
Comment (AlexanderAshitkin) [inline src/main/java/org/apache/maven/buildcache/DefaultNormalizedModelProvider.java:126]: where is it now?

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java [hunk 5]
```diff
@@ -180,9 +182,21 @@ public MavenProjectInput( MavenProject project,
...
+                    LOGGER.debug( "Adding an excludePath from property '{}', values is '{}', path is '{}' ",
+                            propertyName, propertyValue, path );
+                }
+
                 filteredOutPaths.add( Paths.get( properties.getProperty( propertyName ) ) );
             }
         }
+        CacheUtils.debugPrintCollection( LOGGER, filteredOutPaths,
+                "List of excluded paths (checked either by fileName or by startsWith prefix)",
+                "Pah entry" );
```
Comment (AlexanderAshitkin) [inline src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java:199]: 'Pah' entry is a medical term likely :-)

File: src/site/markdown/parameters.md [hunk 6]
```diff
@@ -32,6 +32,7 @@ This documents contains various configuration parameters supported by cache engi
 | `-Dmaven.build.cache.lazyRestore=(true/false)`             | Restore artifacts from remote cache lazily                                                                                               | Performance optimization |
 | `-Dmaven.build.cache.restoreGeneratedSources=(true/false)` | Do not restore generated sources and directly attached files                                                                             | Performance optimization |
 | `-Dmaven.build.cache.alwaysRunPlugins=<list of plugins>`   | Comma seprated list of plugins to always run regardless of cache state. An example: `maven-deploy-plugin:*,maven-install-plugin:install` | Remote cache setup/tuning/troubleshooting |
+| `-Dmaven.build.cache.skipLookup=(true/false)`              | Skip looking up artifacts in caches. Does not affect writing artifacts to caches, disables only reading when set to `true`               | May be used to trigger a forced rebuild when maching artifatcs do exist in caches|
```
Comment (AlexanderAshitkin) [inline src/site/markdown/parameters.md:35]: 👍

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java [hunk 7]
```diff
@@ -819,4 +833,22 @@ public int compare( Path f1, Path f2 )
         }
     }
 
+    public static boolean shouldSkipCacheLookup( MavenProject project )
```
Comment (AlexanderAshitkin) [inline src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java:836]: `skipCacheLookup` or `isSkipCacheLookup` will be better

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java [hunk 8]
```diff
@@ -180,9 +182,21 @@ public MavenProjectInput( MavenProject project,
...
+                String propertyValue = properties.getProperty( propertyName );
+                Path path = Paths.get( propertyValue );
+                filteredOutPaths.add( path );
+                if ( LOGGER.isDebugEnabled() )
+                {
+                    LOGGER.debug( "Adding an excludePath from property '{}', values is '{}', path is '{}' ",
+                            propertyName, propertyValue, path );
+                }
+
                 filteredOutPaths.add( Paths.get( properties.getProperty( propertyName ) ) );
```
[the remark this answers is not in the document]
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java]: @AlexanderAshitkin  thanks, fixed

File: src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java (hunk 3 above)
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java]: @AlexanderAshitkin renamed

File: src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java (hunk 2 above)
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/BuildCacheMojosExecutionStrategy.java]: @AlexanderAshitkin renamed to skipCache

File: src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java [hunk 9]
```diff
@@ -153,37 +154,50 @@ public CacheControllerImpl(
...
 
-            CacheResult localBuild = findLocalBuild( mojoExecutions, context );
+            LOGGER.info( "Attempting to restore project {} from build cache", projectName );
 
-            if ( localBuild.isSuccess() || ( localBuild.isPartialSuccess() && !result.isPartialSuccess() ) )
-            {
-                result = localBuild;
-            }
-            else
+            // remote build first
```
[the remark this answers is not in the document]
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java:177]: Thus is not my comment - just part of code by @gnodet that I've moved inside a block

File: src/site/markdown/parameters.md [hunk 10]
```diff
@@ -32,7 +32,7 @@ This documents contains various configuration parameters supported by cache engi
...
 | `-Dmaven.build.cache.restoreGeneratedSources=(true/false)` | Do not restore generated sources and directly attached files                                                                             | Performance optimization |
 | `-Dmaven.build.cache.alwaysRunPlugins=<list of plugins>`   | Comma seprated list of plugins to always run regardless of cache state. An example: `maven-deploy-plugin:*,maven-install-plugin:install` | Remote cache setup/tuning/troubleshooting |
-| `-Dmaven.build.cache.skipLookup=(true/false)`              | Skip looking up artifacts in caches. Does not affect writing artifacts to caches, disables only reading when set to `true`               | May be used to trigger a forced rebuild when maching artifatcs do exist in caches|
+| `-Dmaven.build.cache.skip=(true/false)`                    | Skip looking up artifacts in caches. Does not affect writing artifacts to caches, disables only reading when set to `true`               | May be used to trigger a forced rebuild when maching artifatcs do exist in caches|
```
Comment (AlexanderAshitkin) [inline src/site/markdown/parameters.md:35]: small typo. maching -> matching

**Review by AlexanderAshitkin — APPROVED** [review]

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java (hunk 5 above)
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java]: @AlexanderAshitkin fixed

File: src/site/markdown/parameters.md (hunk 6 above)
Comment (eltsovalex) [reply src/site/markdown/parameters.md]: @AlexanderAshitkin renamed to -Dorg.maven.build.cache.skipCache

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java (hunk 7 above)
Comment (eltsovalex) [reply src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java]: @AlexanderAshitkin fixed

File: src/site/markdown/parameters.md (hunk 10 above)
Comment (eltsovalex) [reply src/site/markdown/parameters.md]: @AlexanderAshitkin  - thx, fixed

**Comment by jira-importer** [comment]
Resolve #285

**Comment by jira-importer** [comment]
Resolve #285

**Pull request merged**